grpc/server_test.go

```diff
@@ -70,6 +84,190 @@ func TestClientServer(t *testing.T) {
+	}
+
+	if diff := cmp.Diff(receivedFileMatches, mock.SearchResult.ToProto().GetFiles(), protocmp.Transform()); diff != "" {
+		t.Fatalf("unexpected difference in file matches (-want +got):\n%s", diff)
+	}
+}
+
+func TestFuzzGRPCChunkSender(t *testing.T) {
+	validateResult := func(input zoekt.SearchResult) error {
+		// Setup: we constrain the input to be a valid zoekt.SearchResult by ensuring that the
+		// RepoURLs and LineFragment fields are derived from the FileMatches
+		// (instead of being random maps from quick.Check)
+		clientStream, serverStream := newPairedSearchStream(t)
+		sender := gRPCChunkSender(serverStream)
+
+		sender.Send(&input)
+
+		allResponses := readAllStream(t, clientStream)
+		if len(allResponses) == 0 {
+			return errors.New("received no responses")
+		}
+
+		expectedResult := input.ToProto()
+
+		for i, receivedResponse := range allResponses {
+			// First, check some invariants about the progress field
+
+			if i == len(allResponses)-1 {
+				// The last response should have the same progress as the original search result
+				if diff := cmp.Diff(expectedResult.GetProgress(), receivedResponse.GetProgress(), protocmp.Transform()); diff != "" {
+					return fmt.Errorf("unexpected difference in progress (-want +got):\n%s", diff)
+				}
+
+			} else {
+				// All other responses should ensure that the progress' priority is less than the max-pending priority, to
+				// ensure that the client consumes the entire set of chunks
+
+				if receivedResponse.GetProgress().GetPriority() > receivedResponse.GetProgress().GetMaxPendingPriority() {
+					return fmt.Errorf(
+						"received response %d (%s) has priority %.6f, which is greater than the max pending priority %.6f",
+						i, receivedResponse,
+						receivedResponse.GetProgress().GetPriority(), receivedResponse.GetProgress().GetMaxPendingPriority(),
+					)
+
+				}
+			}
+
+			// Safety, ensure that all other fields are echoed back correctly if the schema ever changes
+			opts := []cmp.Option{
+				protocmp.Transform(),
+				protocmp.IgnoreFields(&v1.SearchResponse{},
+					"progress", // progress is tested above
+					"stats",    // aggregated stats are tested below
+					"files",    // files are tested separately
+				),
+			}
+
+			if diff := cmp.Diff(expectedResult, receivedResponse, opts...); diff != "" {
+				return fmt.Errorf("unexpected difference in response fields (-want +got):\n%s", diff)
+			}
+		}
+
+		receivedStats := &zoekt.Stats{}
+
+		var receivedFileMatches []*v1.FileMatch
+		for _, r := range allResponses {
+			receivedStats.Add(zoekt.StatsFromProto(r.GetStats()))
+			receivedFileMatches = append(receivedFileMatches, r.GetFiles()...)
+		}
+
+		// Check to make sure that we get one set of stats back
+		if diff := cmp.Diff(expectedResult.GetStats(), receivedStats.ToProto(),
+			protocmp.Transform(),
+			protocmp.IgnoreFields(&v1.Stats{},
+				"duration", // for whatever the duration field isn't updated when zoekt.Stats.Add is called
```

Comment:
This is irrelevant to the PR itself, but do we know why Duration isn't considered in Add()? 

https://github.com/sourcegraph/zoekt/blob/956d775e32b369708c457dc612fd5b69ba85523e/api.go#L413-L439

Reply:
Duration and Wait are both set once and measure the wall clock time for the whole request. Everything else in here is calculated per shard.